src/ccl_correlation.c

```diff
@@ -506,3 +509,419 @@ void ccl_correlation_3d(ccl_cosmology *cosmo, double a,
 
   return;
 }
+
+/*--------ROUTINE: ccl_correlation_multipole ------
+TASK: Calculate multipole of the redshift space correlation function. Do so using FFTLog.
+
+INPUT:  cosmology, scale factor a, beta (= growth rate / bias), 
+        multipole order l = 0, 2, or 4, number of s values, s values
+
+Multipole function result will be in array xi
+ */
+
+void ccl_correlation_multipole(ccl_cosmology *cosmo, double a, double beta,
+                               int l, int n_s, double *s, double *xi,
+                               int *status) {
+  int i, N_ARR;
+  double *k_arr, *pk_arr, *s_arr, *xi_arr, *xi_arr0;
+
+  N_ARR = (int)(ccl_splines->N_K_3DCOR * log10(ccl_splines->K_MAX / ccl_splines->K_MIN));
+
+  k_arr = ccl_log_spacing(ccl_splines->K_MIN, ccl_splines->K_MAX, N_ARR);
+  if (k_arr == NULL) {
+    *status = CCL_ERROR_MEMORY;
+    strcpy(cosmo->status_message,
+           "ccl_correlation.c: ccl_correlation_multipole ran out of memory\n");
+    return;
+  }
+
+  pk_arr = malloc(N_ARR * sizeof(double));
+  if (pk_arr == NULL) {
+    free(k_arr);
+    *status = CCL_ERROR_MEMORY;
+    strcpy(cosmo->status_message,
+           "ccl_correlation.c: ccl_correlation_multipole ran out of memory\n");
+    return;
+  }
+
+  for (i = 0; i < N_ARR; i++)
+    pk_arr[i] = ccl_nonlin_matter_power(cosmo, k_arr[i], a, status);
```

Comment:
No, no. By default we calculate correlation function for liner power spectrum. Or, for BAO analysis you calculate it separately for wiggle and no-wiggle parts with wiggle part getting LPT damping... but this is beyond current scope. You should make it at minimum optional with default to linear.

Reply:
The CCL function ccl_nonlinear_power will return either the linear or nonlinear power spectrum depending on how the cosmology object (cosmo) is set up by the user. If cosmo is set up with matter_power_spectrum=linear it will use the linear P(k). For the benchmarks we do use the nonlinear P(k), since we want to compare differences in s^2 xi(s) with the BAO peak.

Reply:
So, for BAO, the accurate, well tested model that is used in BAO fitting code is to take linear power spectrum, split it into wiggle and smooth parth and then model non-linearties as sum of smooth correlation function plus zeldovich smoothing with free parameters on the wiggle plart (with sigmas usually marginalised over). For CCL we don't need all this baggage, this is just to say that non-linear xi is not really used in BAO. But fine, as long as the code can also return the linear xi if asked (by configuring cosmology object), I'm fine with that, definitely useful to have several options.